extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -480,6 +508,138 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+    let rtcTotalTime;
+    const disableSWR = (
+        rtcConfig['disableStaleWhileRevalidate'] === true);
+    const headers = new Headers();
+    headers.append('Cache-Control', 'max-age=0');
+    const xhrInit = {credentials: 'include'};
+    const startTime = Date.now();
+    // Because we are wrapping the RTC request in the timeout,
+    // we are guaranteeing that if the RTC is slow to return and
+    // times out for the first slot, it won't be used for any
+    // other slots either. This is in opposition to the other way
+    // we could potentially do it, in which we only still make one
+    // rtc callout, but there is a 1 second timeout from the
+    // perspective of each slot. I.e. if the top slot on the page
+    // calls out for the RTC, which returns after 5 seconds, and
+    // then a slot way down on the page asks for it later.
+    rtcPromise = timerFor(window).timeoutPromise(
+        RTC_TIMEOUT,
+        xhrFor(this.win).fetchJson(
+            endpoint, xhrInit).then(res => {
+              rtcTotalTime = Date.now() - startTime;
+              if (!disableSWR) {
+                // Repopulate the cache.
+                xhrFor(this.win).fetchJson(endpoint, {
+                  credentials: 'include',
+                  headers,
+                }).catch(err => {
+                  user().error(TAG, err.message);
+                });
+              }
+              // Non-200 status codes are forbidden for RTC.
+              // TODO: Add to fetchResponse the ability to
+              // check for redirects as well.
+              if (res.status != 200) {
+                return null;
+              }
+              return res.json().then(rtcResponse => {
+                return {rtcResponse, rtcTotalTime};
+              });
+            }));
+
+    return this.mergeRtc();
+  }
+
+  /**
+   * Merges the RTC response into the jsonTargeting of this.
+   * If it can't merge, or there is no response, potentially
+   * rejects.
+   * @return {Promise} Resolves if ad request is being sent,
+   *     otherwise rejects.
+   */
+  mergeRtc() {
+    // add reasons for promise.reject
+    return rtcPromise.then(
+        r => {
+          // Don't try to merge if we're sending without RTC.
+          let rtcResponse;
+          if (!r || !(rtcResponse = r.rtcResponse)) {
+            return this.shouldSendRequestWithoutRtc(
+                'Bad response');
+          }
+          const rtcTotalTime = r.rtcTotalTime;
+          if (!!rtcResponse['targeting'] ||
+              !!rtcResponse['categoryExclusions']) {
+            this.jsonTargeting_['targeting'] = deepMerge(
+                this.jsonTargeting_['targeting'] || {},
+                rtcResponse['targeting'] || {});
+            this.jsonTargeting_['categoryExclusions'] = deepMerge(
+                this.jsonTargeting_['categoryExclusions'] || {},
+                rtcResponse['categoryExclusions'] || {});
+          }
+          return Promise.resolve(rtcTotalTime);
```

Comment:
Note that total time does not include time for the merge... I suppose that's fine but let's add a comment